### PR TITLE
feat: Improved readme performance & enforced that all readme content but `## Notes` is always created anew

### PR DESCRIPTION
**Pull request by AlexanderSehr**
## Description

- Updated the script so that
  - ReadMes are always regenerated from the ground up 
  - Only the content of a '## Notes' sections is kept and added at the end of the readme file
  - Improved performance by only fetching cross-references one (instead during each template run)
  - Improved usage of variables in threads
  - Updated readme with new header

Re-ran on all current scripts and (as expected) did not trigger any unexpected changes to anything already merged.

| Pipeline |
| - |
| [avm.res.network.private-endpoint](https://github.com/AlexanderSehr/bicep-registry-modules/actions/workflows/avm.res.network.private-endpoint.yml) |

### CODE REVIEW

File: avm/res/key-vault/vault/README.md [hunk 1]
```diff
@@ -1394,8 +1393,4 @@ Resource tags.
 
 ## Cross-referenced modules
 
-This section gives you an overview of all local-referenced module files (i.e., other CARML modules that are referenced in this module) and all remote-referenced files (i.e., Bicep modules that are referenced from a Bicep Registry or Template Specs).
-
-| Reference | Type |
-| :-- | :-- |
-| `br/public:avm-res-network-privateendpoint:0.1.1` | Remote reference |
+_None_
```
Comment (eriqua) [inline avm/res/key-vault/vault/README.md:1396]: Nope :bomb:

Comment (AlexanderSehr) [reply avm/res/key-vault/vault/README.md]: Regen & fixed. Was an issue with the port from an improvement on the CARML-side and the differing paths over there

File: avm/utilities/pipelines/sharedScripts/Set-ModuleReadMe.ps1 [hunk 2]
```diff
@@ -1107,12 +1111,11 @@ function Set-UsageExamplesSection {
 
     # Process content
     $SectionContent = [System.Collections.ArrayList]@(
-        'The following module usage examples are retrieved from the content of the files hosted in the module''s `tests` folder.',
-        '   >**Note**: The name of each example is based on the name of the file from which it is taken.',
+        "The following section provides usage examples for the module, which were used to validate and deploy the module successfully. For a full reference, please review the module's test folder in its repository.",
```
Comment (eriqua) [inline avm/utilities/pipelines/sharedScripts/Set-ModuleReadMe.ps1:1114]:
```suggestion
        "The following section provides usage examples for the module, which were used to validate and deploy the module successfully. For a full reference, please review the module's `tests` folder in its repository.",
```

Comment (eriqua) [reply avm/utilities/pipelines/sharedScripts/Set-ModuleReadMe.ps1:1114]: Agreed offline to keep the actual folder name generic

File: avm/utilities/pipelines/sharedScripts/Set-ModuleReadMe.ps1 [hunk 3]
```diff
@@ -1650,60 +1652,41 @@ function Initialize-ReadMe {
...
-            ($hasTests ? '## Usage examples' : $null),
-            ($hasTests ? '' : $null),
-            '## Parameters',
-            '',
-            '## Outputs',
-            '',
-            '## Cross-referenced modules'
-        ) | Where-Object { $null -ne $_ } # Filter null values
-        $readMeFileContent = $initialContent
+    $inTemplateResourceType = (Get-NestedResourceList $TemplateFileContent).type | Select-Object -Unique | Where-Object {
```
Comment (AlexanderSehr) [inline avm/utilities/pipelines/sharedScripts/Set-ModuleReadMe.ps1:1662]: @eriqua, this is where the 'fix' comes in

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**